File: api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceDao.java
package org.pmiops.workbench.db.dao;

import java.util.List;
import org.pmiops.workbench.db.model.DbCohort;
import org.pmiops.workbench.db.model.DbConceptSet;
import org.pmiops.workbench.db.model.DbUserRecentResource;
import org.springframework.data.repository.CrudRepository;

public interface UserRecentResourceDao extends CrudRepository<DbUserRecentResource, Long> {

  long countUserRecentResourceByUserId(long userId);

  DbUserRecentResource findTopByUserIdOrderByLastAccessDate(long userId);

  DbUserRecentResource findByUserIdAndWorkspaceIdAndCohort(
      long userId, long workspaceId, DbCohort cohort);

  DbUserRecentResource findByUserIdAndWorkspaceIdAndNotebookName(
      long userId, long workspaceId, String notebookPath);

  DbUserRecentResource findByUserIdAndWorkspaceIdAndConceptSet(
      long userId, long workspaceId, DbConceptSet conceptSet);

  List<DbUserRecentResource> findUserRecentResourcesByUserIdOrderByLastAccessDateDesc(long userId);
}

File: api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceDao.kt
package org.pmiops.workbench.db.dao

import org.pmiops.workbench.db.model.Cohort
import org.pmiops.workbench.db.model.ConceptSet
import org.pmiops.workbench.db.model.UserRecentResource
import org.springframework.data.repository.CrudRepository

interface UserRecentResourceDao : CrudRepository<UserRecentResource, Long> {

    fun countUserRecentResourceByUserId(userId: Long): Long

    fun findTopByUserIdOrderByLastAccessDate(userId: Long): UserRecentResource

    fun findByUserIdAndWorkspaceIdAndCohort(
            userId: Long, workspaceId: Long, cohort: Cohort): UserRecentResource

    fun findByUserIdAndWorkspaceIdAndNotebookName(
            userId: Long, workspaceId: Long, notebookPath: String): UserRecentResource

    fun findByUserIdAndWorkspaceIdAndConceptSet(
            userId: Long, workspaceId: Long, conceptSet: ConceptSet): UserRecentResource

    fun findUserRecentResourcesByUserIdOrderByLastAccessDateDesc(userId: Long): List<UserRecentResource>
}
